File: src/spread-calculator.ts
import { ArbitrageOpportunity, PriceData } from "./types";

/**
 * Estimate slippage based on trade size vs available liquidity.
 * Simple linear model: slippage_pct ≈ tradeSize / liquidity * 100.
 */
function estimateSlippage(
  liquidity: number,
  tradeSizeUsd: number
): number {
  if (liquidity <= 0) return 0.01; // assume 1% for unknown liquidity
  return (tradeSizeUsd / liquidity) * 100;
}

/**
 * Calculate confidence score based on liquidity depth and spread size.
 * Returns value between 0 and 1.
 */
function calcConfidence(
  buyLiquidity: number,
  sellLiquidity: number,
  spreadPct: number,
  tradeSizeUsd: number
): number {
  const minLiquidity = Math.min(buyLiquidity, sellLiquidity);
  const liquidityScore = Math.min(1, minLiquidity / (tradeSizeUsd * 20));
  const spreadScore = Math.min(1, spreadPct / 2); // cap at 2% spread
  return parseFloat(((liquidityScore * 0.6 + spreadScore * 0.4)).toFixed(2));
}

export function findArbitrageOpportunities(
  priceData: PriceData,
  tradeSizeUsd: number,
  minSpreadPct: number,
  slippageTolerancePct: number
): ArbitrageOpportunity[] {
  const opportunities: ArbitrageOpportunity[] = [];
  const dexNames = Object.keys(priceData.prices);

  for (let i = 0; i < dexNames.length; i++) {
    for (let j = 0; j < dexNames.length; j++) {
      if (i === j) continue;

      const buyDex = dexNames[i];
      const sellDex = dexNames[j];
      const buyData = priceData.prices[buyDex];
      const sellData = priceData.prices[sellDex];

      // We buy on buyDex (lower price) and sell on sellDex (higher price)
      if (sellData.price <= buyData.price) continue;

      const rawSpreadPct =
        ((sellData.price - buyData.price) / buyData.price) * 100;

      const totalFeesPct = (buyData.fee + sellData.fee) * 100;
      const buySlippage = estimateSlippage(buyData.liquidity, tradeSizeUsd);
      const sellSlippage = estimateSlippage(sellData.liquidity, tradeSizeUsd);
      const totalSlippage = buySlippage + sellSlippage;

      const netSpreadPct = rawSpreadPct - totalFeesPct - totalSlippage;

      if (netSpreadPct < minSpreadPct) continue;

      // Reject if any leg slippage exceeds tolerance
      if (
        buySlippage > slippageTolerancePct ||
        sellSlippage > slippageTolerancePct
      ) {
        continue;
      }

      const estimatedProfitUsd = (netSpreadPct / 100) * tradeSizeUsd;
      const confidence = calcConfidence(
        buyData.liquidity,
        sellData.liquidity,
        netSpreadPct,
        tradeSizeUsd
      );

      opportunities.push({
        pair: priceData.pair,
        buy_dex: buyDex,
        buy_price: buyData.price,
        sell_dex: sellDex,
        sell_price: sellData.price,
        spread_pct: parseFloat(netSpreadPct.toFixed(4)),
        estimated_profit_usd: parseFloat(estimatedProfitUsd.toFixed(2)),
        confidence,
      });
    }
  }

  // Sort by estimated profit descending
  return opportunities.sort(
    (a, b) => b.estimated_profit_usd - a.estimated_profit_usd
  );
}

File: go/internal/spreadcalc/spread_calculator.go
package spreadcalc

import (
	"math"
	"sort"

	"github.com/steuf11/arb-bot/internal/types"
)

// estimateSlippage returns an estimated slippage % using a simple linear model:
// slippage ≈ tradeSize / liquidity * 100.
// When liquidity is unknown (≤ 0) a conservative 1 % is assumed.
func estimateSlippage(liquidity, tradeSizeUSD float64) float64 {
	if liquidity <= 0 {
		return 1.0
	}
	return (tradeSizeUSD / liquidity) * 100
}

// calcConfidence scores an opportunity in [0, 1] based on liquidity depth and spread size.
func calcConfidence(buyLiquidity, sellLiquidity, spreadPct, tradeSizeUSD float64) float64 {
	minLiquidity := math.Min(buyLiquidity, sellLiquidity)
	liquidityScore := math.Min(1.0, minLiquidity/(tradeSizeUSD*20))
	spreadScore := math.Min(1.0, spreadPct/2) // cap at 2 % spread
	raw := liquidityScore*0.6 + spreadScore*0.4
	// Round to 2 decimal places.
	return math.Round(raw*100) / 100
}

// roundTo4 rounds v to 4 decimal places.
func roundTo4(v float64) float64 { return math.Round(v*10000) / 10000 }

// roundTo2 rounds v to 2 decimal places.
func roundTo2(v float64) float64 { return math.Round(v*100) / 100 }

// FindArbitrageOpportunities scans priceData for cross-DEX arbitrage opportunities.
// It deducts fees and estimated slippage from the raw spread, filters out pairs
// below minSpreadPct or where any leg exceeds slippageTolerancePct, and returns
// results sorted by estimated profit descending.
func FindArbitrageOpportunities(
	priceData types.PriceData,
	tradeSizeUSD,
	minSpreadPct,
	slippageTolerancePct float64,
) []types.ArbitrageOpportunity {
	var opportunities []types.ArbitrageOpportunity

	dexNames := make([]string, 0, len(priceData.Prices))
	for name := range priceData.Prices {
		dexNames = append(dexNames, name)
	}

	for i := range dexNames {
		for j := range dexNames {
			if i == j {
				continue
			}

			buyDex := dexNames[i]
			sellDex := dexNames[j]
			buyData := priceData.Prices[buyDex]
			sellData := priceData.Prices[sellDex]

			// We buy on buyDex (lower price) and sell on sellDex (higher price).
			if sellData.Price <= buyData.Price {
				continue
			}

			rawSpreadPct := ((sellData.Price - buyData.Price) / buyData.Price) * 100
			totalFeesPct := (buyData.Fee + sellData.Fee) * 100
			buySlippage := estimateSlippage(buyData.Liquidity, tradeSizeUSD)
			sellSlippage := estimateSlippage(sellData.Liquidity, tradeSizeUSD)
			totalSlippage := buySlippage + sellSlippage

			netSpreadPct := rawSpreadPct - totalFeesPct - totalSlippage

			if netSpreadPct < minSpreadPct {
				continue
			}

			// Reject if any leg's slippage exceeds the configured tolerance.
			if buySlippage > slippageTolerancePct || sellSlippage > slippageTolerancePct {
				continue
			}

			estimatedProfitUSD := (netSpreadPct / 100) * tradeSizeUSD
			confidence := calcConfidence(buyData.Liquidity, sellData.Liquidity, netSpreadPct, tradeSizeUSD)

			opportunities = append(opportunities, types.ArbitrageOpportunity{
				Pair:               priceData.Pair,
				BuyDex:             buyDex,
				BuyPrice:           buyData.Price,
				SellDex:            sellDex,
				SellPrice:          sellData.Price,
				SpreadPct:          roundTo4(netSpreadPct),
				EstimatedProfitUSD: roundTo2(estimatedProfitUSD),
				Confidence:         confidence,
			})
		}
	}

	// Sort by estimated profit descending.
	sort.Slice(opportunities, func(i, j int) bool {
		return opportunities[i].EstimatedProfitUSD > opportunities[j].EstimatedProfitUSD
	})

	return opportunities
}
